src/btreemap/node.rs

```diff
@@ -147,34 +152,55 @@ impl<K: Storable + Ord + Clone> Node<K> {
 
     /// Returns true if the node cannot store anymore entries, false otherwise.
     pub fn is_full(&self) -> bool {
-        self.keys.len() >= CAPACITY
+        self.keys_encoded_values.len() >= CAPACITY
     }
 
     /// Swaps the entry at index `idx` with the given entry, returning the old entry.
     pub fn swap_entry<M: Memory>(
         &mut self,
         idx: usize,
-        (mut key, value): Entry<K>,
+        (key, value): Entry<K>,
         memory: &M,
     ) -> Entry<K> {
-        core::mem::swap(&mut self.keys[idx], &mut key);
-        let old_value = self.value(idx, memory).to_vec();
-        self.encoded_values.borrow_mut()[idx] = Value::ByVal(value);
-        (key, old_value)
+        let mut swap_with = (key, RefCell::new(Value::ByVal(value)));
+        core::mem::swap(&mut self.keys_encoded_values[idx], &mut swap_with);
+        (
+            swap_with.0,
+            self.resolve_value(RefCell::into_inner(swap_with.1), memory),
+        )
     }
 
     /// Returns a copy of the entry at the specified index.
     pub fn entry<M: Memory>(&self, idx: usize, memory: &M) -> Entry<K> {
-        (self.keys[idx].clone(), self.value(idx, memory).to_vec())
+        (
+            self.keys_encoded_values[idx].0.clone(),
+            self.value(idx, memory).to_vec(),
+        )
     }
 
     /// Returns a reference to the encoded value at the specified index.
     pub fn value<M: Memory>(&self, idx: usize, memory: &M) -> Ref<[u8]> {
         // Load and cache the value from the underlying memory if needed.
-        {
-            let mut values = self.encoded_values.borrow_mut();
+        let mut value = self.keys_encoded_values[idx].1.borrow_mut();
+        if let Value::ByRef(offset) = *value {
+            // Cache the value internally.
+            *value = Value::ByVal(self.resolve_value(Value::ByRef(offset), memory));
+        }
+        drop(value); // drop borrow because it's reborrowed below
```

Comment:
nit: Could we not use an explicit scope to help the borrow checker? I find that a bit more readable than an explicit `drop` that you need to accompany with the comment. The explicit scope is almost always an indication that you want to force something to be dropped. So, something like:

```
{
   let mut value = self.keys_encoded_values[idx].1.borrow_mut();
   if let Value::ByRef(offset) = *value {
      // Cache the value internally.
      *value = Value::ByVal(self.resolve_value(Value::ByRef(offset), memory));
   }
}
```

Reply:
Actually, I realized that both the old and new implementation had a bug which made the code panic when we called `.value()` multiple times with the same index. I fixed it and added a test, PTAL.